src/neo4j_genai/generation/types.py

```diff
@@ -0,0 +1,43 @@
+#  Copyright (c) "Neo4j"
+#  Neo4j Sweden AB [https://neo4j.com]
+#  #
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#  #
+#      https://www.apache.org/licenses/LICENSE-2.0
+#  #
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+from typing import Any
+from pydantic import BaseModel, ConfigDict
+
+from neo4j_genai.llm import LLMInterface
+from neo4j_genai.generation.prompts import RagTemplate
+from neo4j_genai.retrievers.base import Retriever
+from neo4j_genai.types import RetrieverResult
+
+
+class RagInitModel(BaseModel):
+    retriever: Retriever
+    llm: LLMInterface
+    prompt_template: RagTemplate
+
+    model_config = ConfigDict(arbitrary_types_allowed=True)
+
+
+class RagSearchModel(BaseModel):
+    query: str
+    examples: str = ""
+    retriever_config: dict[str, Any] = {}
```

Comment:
Is there a way to avoid this mutable default?

Reply:
It is handled properly by Pydantic: https://docs.pydantic.dev/latest/concepts/models/#fields-with-non-hashable-default-values

But if you prefer we can use Optional.

Reply:
OK that works 👍